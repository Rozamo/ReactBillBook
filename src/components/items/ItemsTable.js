import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import loadingGif from '../../images/loading.gif';
import LoadData from '../helper/api/LoadData';
import BlueButton from '../bluebutton/BlueButton';
import Table from '../helper/table/Table';

export default function ItemsTable() {
  const [error, setError] = useState(null);
  const [isLoaded, setIsLoaded] = useState(false);
  const [items, setItems] = useState([]);

  async function getData() {
    const data = await LoadData("items");
    if (data && data.entity === 'collection') {
      setIsLoaded(true);
      setItems(data.items);
    }
    else if (data) {
      setError(data);
      setIsLoaded(true);
    }
  }

  useEffect(() => {
    getData();
  }, []);

  function handleContent() {
    if (error)
      return <div>Error: {error.message}</div>;
    else if (!isLoaded)
      return <img src={loadingGif} alt="Loading...." id="load-img"></img>;
    else {
      const path = window.location.pathname;
      const submitSuccess = path.includes('success') ? true : false;
      return Table(items, ['NAME', 'DESCRIPTION', 'PRICE', 'ADDED ON'], ['name', 'description', 'amount', 'created_at'], submitSuccess);
    }
  }
  
  return (
    <div className="content">
      <div className="top-panel">
        <h1 id="title">Items</h1>
        <Link to='/items/create'>
          <BlueButton sidebarChoice="items" contentChoice="list" />
        </Link>
      </div>
      {handleContent()}
    </div>
  );
}