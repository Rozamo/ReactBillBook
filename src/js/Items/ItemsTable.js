import React from 'react';
import load_gif from '../../images/load.gif';
import LoadData from '../Helper/API/LoadData';
import TableHead from '../Helper/TableMaker/TableHead';
import TableBody from '../Helper/TableMaker/TableBody';

export default class ItemsTable extends React.Component {
    constructor() {
        super();
        this.state = {
            error: null,
            isLoaded: false,
            items: []
        };
    }
    async getData() {
        const data = await LoadData(this.props.sidebar_choice);
        if (data && data.entity === 'collection') {
            this.setState({
                isLoaded: true,
                items: data.items
            });
        }
        else if (data) {
            this.setState({
                error: data,
                isLoaded: true
            });
        }
    }
    componentDidMount() {
        this.getData();
    }
    render() {
        const { error, isLoaded, items } = this.state;
        if (error)
            return <div>Error: {error.message}</div>;
        else if (!isLoaded)
            return <img src={load_gif} alt="Loading...." id="load-img"></img>;
        else return <table className="inv-table" id="inv-table">
            {TableHead('NAME', 'DESCRIPTION', 'PRICE', 'ADDED ON')}
            {TableBody(items, 'name', 'description', 'amount', 'created_at')}
        </table>;
    }
}