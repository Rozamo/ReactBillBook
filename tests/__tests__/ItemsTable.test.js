import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import ItemsTable from "../../src/components/items/ItemsTable";

const fakeData = {
  entity: "collection",
  items: [
    {
      name: "Jmp",
      email: "J@rzp.com",
      age: "32",
      address: "123, Charming Avenue",
      amount: 1231,
    },
  ],
};

test("renders ItemsTable", async () => {
  global.fetch = jest.fn().mockImplementation(() =>
    Promise.resolve({
      json: () => new Promise((resolve) => resolve(fakeData)),
    })
  );

  const { container } = render(
    <Router>
      <ItemsTable location={{ state: { submitSuccess: true } }} />
    </Router>
  );

  expect(container.querySelector("#load-img").getAttribute("alt")).toBe(
    "Loading...."
  );

  expect(container.querySelector(".content")).toMatchSnapshot();

  await waitFor(() => screen);

  expect(container.querySelector(".content")).toMatchSnapshot();

  expect(fetch).toHaveBeenCalledTimes(1);
  expect(container.querySelector("tbody").textContent).toContain("Jmp");

  global.fetch.mockClear();
  delete global.fetch;
});

test("handles error ItemsTable", async () => {
  global.fetch = jest.fn().mockImplementation(() =>
    Promise.resolve({
      ok: false,
      json: () => new Promise((resolve) => resolve({ status: 500 })),
    })
  );

  const { container } = render(
    <Router>
      <ItemsTable location={{ state: { submitSuccess: true } }} />
    </Router>
  );
  expect(container.querySelector("#load-img").getAttribute("alt")).toBe(
    "Loading...."
  );

  expect(container).toMatchSnapshot();

  await waitFor(() => screen);

  expect(container.querySelector(".content")).toMatchSnapshot();

  global.fetch.mockClear();
  delete global.fetch;
});
