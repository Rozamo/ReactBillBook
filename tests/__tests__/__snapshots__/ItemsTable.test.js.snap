// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles error ItemsTable 1`] = `
<div>
  <div
    class="content"
  >
    <div
      class="top-panel"
    >
      <h1
        id="title"
      >
        Items
      </h1>
      <a
        href="/items/create"
      >
        <button
          id="button"
          type="button"
          value=""
        >
          <img
            alt="New"
            id="floppy"
            src="test-file-stub"
          />
          Add Item
        </button>
      </a>
    </div>
    <img
      alt="Loading...."
      id="load-img"
      src="test-file-stub"
    />
  </div>
</div>
`;

exports[`handles error ItemsTable 2`] = `
<div
  class="content"
>
  <div
    class="top-panel"
  >
    <h1
      id="title"
    >
      Items
    </h1>
    <a
      href="/items/create"
    >
      <button
        id="button"
        type="button"
        value=""
      >
        <img
          alt="New"
          id="floppy"
          src="test-file-stub"
        />
        Add Item
      </button>
    </a>
  </div>
  <div>
    Error:
  </div>
</div>
`;

exports[`renders ItemsTable 1`] = `
<div
  class="content"
>
  <div
    class="top-panel"
  >
    <h1
      id="title"
    >
      Items
    </h1>
    <a
      href="/items/create"
    >
      <button
        id="button"
        type="button"
        value=""
      >
        <img
          alt="New"
          id="floppy"
          src="test-file-stub"
        />
        Add Item
      </button>
    </a>
  </div>
  <img
    alt="Loading...."
    id="load-img"
    src="test-file-stub"
  />
</div>
`;

exports[`renders ItemsTable 2`] = `
<div
  class="content"
>
  <div
    class="top-panel"
  >
    <h1
      id="title"
    >
      Items
    </h1>
    <a
      href="/items/create"
    >
      <button
        id="button"
        type="button"
        value=""
      >
        <img
          alt="New"
          id="floppy"
          src="test-file-stub"
        />
        Add Item
      </button>
    </a>
  </div>
  <table
    class="inv-table"
    id="inv-table"
  >
    <thead>
      <tr>
        <th>
          NAME
        </th>
        <th>
          DESCRIPTION
        </th>
        <th>
          PRICE
        </th>
        <th>
          ADDED ON
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="newTD firstTD"
        >
          Jmp
        </td>
        <td
          class="newTD middleTD"
        />
        <td
          class="newTD middleTD"
        >
          â‚¹12.31
        </td>
        <td
          class="newTD lastTD"
        >
          NaN Invalid Date NaN
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
